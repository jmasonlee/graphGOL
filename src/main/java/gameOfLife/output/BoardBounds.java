package gameOfLife.output;

import gameOfLife.cell.Cell;

import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.function.Function;

public class BoardBounds {
  public static final int DEFAULT_BOARD_SIZE = 5;
  private int upperYValue = 5;
  private int leftmostXValue = 0;
  private int height = DEFAULT_BOARD_SIZE;
  private int width = DEFAULT_BOARD_SIZE;

  public BoardBounds(List<Cell> cellsForBoard) {
    calculateBoardBounds(cellsForBoard);
  }

  public int getWidth() {
    return width;
  }

  public int getHeight() {
    return height;
  }

  public int getUpperYValue() {
    return upperYValue;
  }

  public int getLeftmostXValue() {
    return leftmostXValue;
  }

  private void calculateBoardBounds(List<Cell> cellsForBoard) {
    if (cellsForBoard.isEmpty()) {
      return;
    }

    Function<Cell, Integer> fetchYvalue = c->c.y;
    upperYValue = getCellWithMaxCoordinate(cellsForBoard, fetchYvalue).y;
    int lowerYValue = getCellWithLowestCoordinate(cellsForBoard, fetchYvalue).y;
    height = getDimension(lowerYValue, upperYValue);

    Function<Cell, Integer> fetchXvalue = c->c.x;
    leftmostXValue = getCellWithLowestCoordinate(cellsForBoard, fetchXvalue).x;
    int rightmostXValue = getCellWithMaxCoordinate(cellsForBoard, fetchXvalue).x;
    width = getDimension(rightmostXValue, leftmostXValue);
  }

  private Cell getCellWithLowestCoordinate(List<Cell> cellsForBoard, Function<Cell, Integer> fetchCoordinate) {
    return Collections.min(cellsForBoard, Comparator.comparing(fetchCoordinate));
  }

  private Cell getCellWithMaxCoordinate(List<Cell> cellsForBoard, Function<Cell, Integer> fetchCoordinate) {
    return Collections.max(cellsForBoard, Comparator.comparing(fetchCoordinate));
  }

  private int getDimension(int lowestDimensionValue, int highestDimensionValue) {
    return Math.abs(highestDimensionValue - lowestDimensionValue) < DEFAULT_BOARD_SIZE ?
        DEFAULT_BOARD_SIZE :
        calculateDimensionLargerThanDefault(lowestDimensionValue, highestDimensionValue);
  }

  private int calculateDimensionLargerThanDefault(int lowestDimensionValue, int highestDimensionValue) {
    return Math.abs(highestDimensionValue - lowestDimensionValue) + 1;
  }

}
