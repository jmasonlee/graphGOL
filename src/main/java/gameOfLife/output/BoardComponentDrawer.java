package gameOfLife.output;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class BoardComponentDrawer {
  static final String TOP_BOUNDARY_END = " \n";
  static final String CELL = "|__";
  static final String ROW_END = "|\n";

  static List<String> emptyBoardOfSize(BoardBounds boardBounds) {
    List<String> board = new ArrayList<String>();

    Coordinates coordinates = new Coordinates(boardBounds);

    board.addAll(createHeaderRow(coordinates.xCoordinates, coordinates.widestCoordinateString));
    board.addAll(createEmptyRows(boardBounds, coordinates.yCoordinates));

    return board;
  }

  static List<String> createHeaderRow(List<String> xCoords, Integer widestY) {
    List<String> board = new ArrayList<String>();

    board.add(" " + String.join("", Collections.nCopies(widestY, " ")));
    board.addAll(formatIndividualColumnHeaders(xCoords));
    board.add(TOP_BOUNDARY_END);

    return board;
  }

  static List<String> formatIndividualColumnHeaders(List<String> xCoords) {
    List<String> formattedCoords = new ArrayList<String>();
    formattedCoords.add(xCoords.get(0));

    for (int i = 1; i < xCoords.size(); i++) {
      formattedCoords.add(" " + xCoords.get(i));
    }

    return formattedCoords;
  }

  static List<String> createEmptyRows(BoardBounds boardBounds, List<String> yCoords) {
    List<String> board = new ArrayList<String>();

    for (int i = 0; i < boardBounds.getHeight(); i++) {
      board.add(yCoords.get(i));
      board.addAll(Collections.nCopies(boardBounds.getWidth(), CELL));
      board.add(ROW_END);
    }

    return board;
  }
}