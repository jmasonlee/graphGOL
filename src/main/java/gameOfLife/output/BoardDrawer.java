package gameOfLife.output;

import gameOfLife.cell.Cell;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

public class BoardDrawer {

  static List<String> drawHeaderRow(Coordinates coordinates) {
    List<String> board = new ArrayList<>();

    board.add(String.join("", Collections.nCopies(coordinates.widestCoordinateString, BoardSquare.SPACE.getValue())));
    board.addAll(addSpacingToColumnHeaders(coordinates.xCoordinates));
    board.add(BoardSquare.TOP_BOUNDARY_END.getValue());

    return board;
  }

  private static List<String> addSpacingToColumnHeaders(List<String> headers) {
    return headers.stream().map(c -> BoardSquare.SPACE.getValue() + c).collect(Collectors.toList());
  }

  static List<String> drawLiveCellsOnBoard(List<String> oldBoard, List<Cell> cells, BoardBounds boardBounds) {
    cells.forEach(c -> {
      int positionInBoard = getCellPositionInBoard(c, boardBounds);
      oldBoard.set(positionInBoard, BoardSquare.LIVE_CELL.getValue());
    });

    return oldBoard;
  }

  private static int getCellPositionInBoard(Cell cell, BoardBounds boardBounds) {
    int adjustedX = Math.abs(boardBounds.getLeftmostXValue() - cell.x);
    int adjustedY = Math.abs(boardBounds.getUpperYValue() - cell.y);

    int boardWidth = 2 + boardBounds.getWidth();

    return boardWidth + (adjustedY * boardWidth) + (adjustedX + 1);
  }
}
