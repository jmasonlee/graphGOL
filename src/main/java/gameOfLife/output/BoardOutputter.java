package gameOfLife.output;

import gameOfLife.cell.Cell;
import org.apache.commons.lang.StringUtils;

import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.function.Function;

public class BoardOutputter {

  private static final String TOP_BOUNDARY = " __";

  public static String createBoardOutput(List<Cell> cells) {
    BoardBounds boardBounds = new BoardBounds(cells);
    StringBuilder emptyBoard = emptyBoardOfSize(boardBounds.getWidth(), boardBounds.getHeight());
    return populateBoardWithCells(emptyBoard, cells, boardBounds);
  }

  private static String populateBoardWithCells(StringBuilder board, List<Cell> cells, BoardBounds boardBounds) {
    cells.forEach(c -> {
      int positionInBoard = getCellPositionInBoard(c, boardBounds);
      board.setCharAt(positionInBoard, 'X');
    });
    return board.toString();
  }

  private static int getCellPositionInBoard(Cell cell, BoardBounds boardBounds) {
    int adjustedX = Math.abs(boardBounds.getLeftmostXValue() - cell.x);
    int adjustedY = Math.abs(boardBounds.getUpperYValue() - cell.y);

    int boardWidth = (3 * boardBounds.getWidth()) + 2;

    return boardWidth + (adjustedY * boardWidth) + (adjustedX * 3);
  }

  private static StringBuilder emptyBoardOfSize(int width, int height) {
    StringBuilder board = new StringBuilder();

    board.append(StringUtils.repeat(TOP_BOUNDARY, width));
    board.append("\n");

    for (int i = 0; i < height; i++) {
      board.append(StringUtils.repeat("|__", width));
      board.append("|\n");
    }

    return board;
  }

  public static class BoardBounds {
    public static final int DEFAULT_BOARD_SIZE = 5;
    private int upperYValue = 5;
    private int leftmostXValue = 0;
    private int height = DEFAULT_BOARD_SIZE;
    private int width = DEFAULT_BOARD_SIZE;

    public BoardBounds(List<Cell> cellsForBoard) {
      calculateBoardBounds(cellsForBoard);
    }

    public int getWidth() {
      return width;
    }

    public int getHeight() {
      return height;
    }

    public int getUpperYValue() {
      return upperYValue;
    }

    public int getLeftmostXValue() {
      return leftmostXValue;
    }

    private void calculateBoardBounds(List<Cell> cellsForBoard) {
      if (cellsForBoard.isEmpty()) {
        return;
      }

      Function<Cell, Integer> fetchYvalue = c->c.y;
      upperYValue = getCellWithMaxDimension(cellsForBoard, fetchYvalue).y;
      int lowerYValue = getCellWithMinDimension(cellsForBoard, fetchYvalue).y;
      height = getDimension(lowerYValue, upperYValue);

      Function<Cell, Integer> fetchXvalue = c->c.x;
      leftmostXValue = getCellWithMinDimension(cellsForBoard, fetchXvalue).x;
      int rightmostXValue = getCellWithMaxDimension(cellsForBoard, fetchXvalue).x;
      width = getDimension(rightmostXValue, leftmostXValue);
    }

    private Cell getCellWithMinDimension(List<Cell> cellsForBoard, Function<Cell, Integer> fetchCellDimension) {
      return Collections.min(cellsForBoard, Comparator.comparing(c -> fetchCellDimension.apply(c)));
    }

    private Cell getCellWithMaxDimension(List<Cell> cellsForBoard, Function<Cell, Integer> fetchCellDimension) {
      return Collections.max(cellsForBoard, Comparator.comparing(c->fetchCellDimension.apply(c)));
    }

    private int getDimension(int lowestDimensionValue, int highestDimensionValue) {
      return Math.abs(highestDimensionValue - lowestDimensionValue) < DEFAULT_BOARD_SIZE ?
          DEFAULT_BOARD_SIZE :
          calculateDimensionLargerThanDefault(lowestDimensionValue, highestDimensionValue);
    }

    private int calculateDimensionLargerThanDefault(int lowestDimensionValue, int highestDimensionValue) {
      return Math.abs(highestDimensionValue - lowestDimensionValue) + 1;
    }

  }
}