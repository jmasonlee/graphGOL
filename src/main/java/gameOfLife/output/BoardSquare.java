package gameOfLife.output;

public enum BoardSquare {
  TOP_BOUNDARY_END(" \n"),
  EMPTY_CELL("|__"),
  ROW_END("|\n"),
  SPACE(" "),
  LIVE_CELL("|X_");

  private String value;
  public int MIN_WIDTH = 2;

  public String getValue() {
    return value;
  }

  BoardSquare(String value) {
    this.value = value;
  }
}
