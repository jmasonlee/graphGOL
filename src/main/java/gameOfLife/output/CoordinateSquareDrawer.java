package gameOfLife.output;

public class CoordinateSquareDrawer {
  protected static String drawCoordinateSquare(String yCoordinate) {
    String coordinateSquare = yCoordinate;
    if (yCoordinate.length() == 1) {
      coordinateSquare = BoardSquare.SPACE.getValue() + coordinateSquare ;
    }
    return coordinateSquare;
  }
}
