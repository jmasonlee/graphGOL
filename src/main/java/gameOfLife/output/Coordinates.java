package gameOfLife.output;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class Coordinates {
  List<String> xCoordinates;
  List<String> yCoordinates;
  int widestCoordinateString;

  public Coordinates(List<String> xCoordinates, List<String> yCoordinates) {
    this.xCoordinates = xCoordinates;
    this.yCoordinates = yCoordinates;
    this.widestCoordinateString = getLargestNumberOfCharsInCoordinate();
  }

  static List<String> createXCoordinates(int leftmostXValue, int width) {
    int endAt = leftmostXValue + width -1;
    return createCoordinates(leftmostXValue, endAt);
  }

  static List<String> createYCoordinates(int upperYValue, int dimensionSize) {
    int startAt = upperYValue - dimensionSize + 1;

    List<String> coordinateNumbers = createCoordinates(startAt, upperYValue);

    Collections.reverse(coordinateNumbers);

    return coordinateNumbers;
  }

  private static List<String> createCoordinates(int startAt, int endAt) {
    return IntStream.rangeClosed(startAt, endAt).boxed()
      .map(coord -> coord < 0 || 9 < coord ? coord.toString() : " " + coord.toString())
      .collect(Collectors.toList());
  }

  private int getLargestNumberOfCharsInCoordinate() {
    List<String> allCoordinateStrings = new ArrayList<>(xCoordinates);
    allCoordinateStrings.addAll(yCoordinates);
    return Collections.max(allCoordinateStrings, Comparator.comparing(String::length)).length();
  }

  @Override
  public String toString() {
    return "Coordinates{" +
      "xCoordinates=" + xCoordinates +
      ", yCoordinates=" + yCoordinates +
      ", widestCoordinateString=" + widestCoordinateString +
      '}';
  }
}
