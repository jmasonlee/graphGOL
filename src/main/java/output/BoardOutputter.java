package output;

import cell.Cell;
import org.apache.commons.lang.StringUtils;

import java.util.Collections;
import java.util.Comparator;
import java.util.List;

public class BoardOutputter {

  public static String createBoardOutput(List<Cell> cells) {
    BoardBounds boardBounds = new BoardBounds(cells);
    StringBuilder emptyBoard = emptyBoardOfSize(boardBounds.getWidth(), boardBounds.getHeight());
    return populateBoardWithCells(emptyBoard, cells, boardBounds);
  }

  private static String populateBoardWithCells(StringBuilder board, List<Cell> cells, BoardBounds boardBounds) {
    cells.forEach(c -> {
      int positionInBoard = getCellPositionInBoard(c, boardBounds);
      board.setCharAt(positionInBoard, 'X');
    });
    return board.toString();
  }

  private static int getCellPositionInBoard(Cell cell, BoardBounds boardBounds) {
    int adjustedX = Math.abs(boardBounds.getLeftmostXValue() - cell.x);
    int adjustedY = Math.abs(boardBounds.getUpperYValue() - cell.y);
    return (adjustedY * boardBounds.getWidth()) + adjustedY + adjustedX;
  }

  private static StringBuilder emptyBoardOfSize(int width, int height) {
    StringBuilder board = new StringBuilder();

    for (int i = 0; i < height; i++) {
      board.append(StringUtils.repeat(" ", width));
      if (i < height - 1) {
        board.append("\n");
      }
    }

    return board;
  }

  public static class BoardBounds {
    public static final int DEFAULT_BOARD_SIZE = 5;
    private int upperYValue = 5;
    private int leftmostXValue = 0;
    private int height = DEFAULT_BOARD_SIZE;
    private int width = DEFAULT_BOARD_SIZE;

    public BoardBounds(List<Cell> cellsForBoard) {
      calculateBoardBounds(cellsForBoard);
    }

    public int getWidth() {
      return width;
    }

    public int getHeight() {
      return height;
    }

    public int getUpperYValue() {
      return upperYValue;
    }

    public int getLeftmostXValue() {
      return leftmostXValue;
    }

    private void calculateBoardBounds(List<Cell> cellsForBoard) {
      if (cellsForBoard.isEmpty()) {
        return;
      }

      upperYValue = Collections.max(cellsForBoard, Comparator.comparing(cell -> cell.y)).y;
      int lowerYValue = Collections.min(cellsForBoard, Comparator.comparing(cell -> cell.y)).y;
      height =
          getDimension(lowerYValue, upperYValue, this.height);

      leftmostXValue = Collections.min(cellsForBoard, Comparator.comparing(cell -> cell.x)).x;
      int rightmostXValue = Collections.max(cellsForBoard, Comparator.comparing(cell -> cell.x)).x;
      width =
          getDimension(rightmostXValue, leftmostXValue, this.width);
    }

    private int getDimension(int lowestDimensionValue, int highestDimensionValue, int p) {
      return Math.abs(highestDimensionValue - lowestDimensionValue) < DEFAULT_BOARD_SIZE ?
          p :
          Math.abs(highestDimensionValue - lowestDimensionValue) + 1;
    }

  }
}