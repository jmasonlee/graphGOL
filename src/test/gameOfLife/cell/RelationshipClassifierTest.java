package gameOfLife.cell;

import gameOfLife.TestUtils.CellCoverage;
import gameOfLife.output.BoardOutputter;
import org.approvaltests.Approvals;
import org.junit.Test;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class RelationshipClassifierTest {

  @Test
  public void testCellClassification() {
    List<Cell> cellsToClassify = getListOfCellsToClassify(new Cell(0, 0));

    Map<Relationships, List<Cell>> relationship =
        RelationshipClassifier.classify(new Cell(0, 0), cellsToClassify).getRelationships();

    List<String> relationshipBoards = new ArrayList<>();
    relationshipBoards.add(
        "\n" + BoardOutputter.createBoardOutput(relationship.get(Relationships.SELF)));
    relationshipBoards.add(
        "\n" + BoardOutputter.createBoardOutput(relationship.get(Relationships.NEIGHBOUR)));
    relationshipBoards.add(
        "\n" + BoardOutputter.createBoardOutput(relationship.get(Relationships.COPARENT)));
    relationshipBoards.add(
        "\n" + BoardOutputter.createBoardOutput(relationship.get(Relationships.DISCONNECTED)));
    Approvals.verifyAll("Boards", relationshipBoards);
  }

  public List<Cell> getListOfCellsToClassify(Cell cell) {
    int maxOneDimensionalDistance = 3;

    Cell lowerLeftCell =
        new Cell(
            relativeLowerLeftLocaltion(maxOneDimensionalDistance, cell.x),
            relativeLowerLeftLocaltion(maxOneDimensionalDistance, cell.y));
    Cell upperRightCell =
        new Cell(
            relativeUpperRightLocation(maxOneDimensionalDistance, cell.x),
            relativeUpperRightLocation(maxOneDimensionalDistance, cell.y));
    List<Cell> cellsToClassify =
        CellCoverage.generateAllPossibleCellsBetweenTwoCells(lowerLeftCell, upperRightCell);

    return cellsToClassify;
  }

  private int relativeUpperRightLocation(int maxOneDimensionalDistance, Integer centralLocation) {
    return centralLocation + maxOneDimensionalDistance + 1;
  }

  private int relativeLowerLeftLocaltion(int maxOneDimensionalDistance, Integer centralLocation) {
    return centralLocation - maxOneDimensionalDistance;
  }
}
