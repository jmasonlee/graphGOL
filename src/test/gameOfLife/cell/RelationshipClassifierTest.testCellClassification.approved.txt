Boards[0] = 
 __ __ __ __ __ 
|X_|__|__|__|__|
|__|__|__|__|__|
|__|__|__|__|__|
|__|__|__|__|__|
|__|__|__|__|__|

Boards[1] = 
 __ __ __ __ __ 
|X_|X_|X_|__|__|
|X_|__|X_|__|__|
|X_|X_|X_|__|__|
|__|__|__|__|__|
|__|__|__|__|__|

Boards[2] = 
 __ __ __ __ __ 
|X_|X_|X_|X_|X_|
|X_|__|__|__|X_|
|X_|__|__|__|X_|
|X_|__|__|__|X_|
|X_|X_|X_|X_|X_|

Boards[3] = 
 __ __ __ __ __ __ __ __ 
|X_|X_|X_|X_|X_|X_|X_|X_|
|X_|X_|X_|X_|X_|X_|X_|X_|
|X_|__|__|__|__|__|X_|X_|
|X_|__|__|__|__|__|X_|X_|
|X_|__|__|__|__|__|X_|X_|
|X_|__|__|__|__|__|X_|X_|
|X_|__|__|__|__|__|X_|X_|
|X_|X_|X_|X_|X_|X_|X_|X_|

